File: sdks/dotnet/sulfone-helium/Domain/Extension/Service.cs
using sulfone_helium.Domain.Core;
using sulfone_helium.Domain.Service;

namespace sulfone_helium.Domain.Extension;

public class ExtensionService
{
    public required ICyanExtension Ext { private get; init; }

    public async Task<IExtensionOutput> Extend(ExtensionAnswerInput answer)
    {
        short pointer = -1;
        var i = new StatelessInquirer(answer.Answers, ref pointer);
        var d = new StatelessDeterminism(answer.DeterministicState, ref pointer);
        try
        {
            var input = new CyanExtensionInput
            {
                PrevAnswers = answer.PrevAnswers,
                PrevCyan = answer.PrevCyan,
                PrevExtensionAnswers = answer.PrevExtensionAnswers,
                PrevExtensionCyans = answer.PrevExtensionCyans,
            };
            var r = await Ext.Extension(i, d, input);
            return new ExtensionFinalOutput(r);
        }
        catch (OutOfAnswerException e)
        {
            return new ExtensionQnAOutput(d.States, e.Question);
        }
    }

    public async Task<string?> Validate(ExtensionValidateInput answer)
    {
        short pointer = -1;
        var i = new StatelessInquirer(answer.Answers, ref pointer);
        var d = new StatelessDeterminism(answer.DeterministicState, ref pointer);
        try
        {
            var input = new CyanExtensionInput()
            {
                PrevAnswers = answer.PrevAnswers,
                PrevCyan = answer.PrevCyan,
                PrevExtensionAnswers = answer.PrevExtensionAnswers,
                PrevExtensionCyans = answer.PrevExtensionCyans,
            };
            var r = await Ext.Extension(i, d, input);
            throw new ApplicationException("Not supposed to reach here for validation!");
        }
        catch (OutOfAnswerException e)
        {
            var q = e.Question;
            return q.Validate?.Invoke(answer.Validate);
        }
    }
}

File: sdks/node/src/domain/template/service.ts
import type { ICyanTemplate } from '../core/cyan_script.js';
import type { TemplateInput, TemplateValidateInput } from './input.js';
import type { TemplateFinalOutput, TemplateOutput, TemplateQnAOutput } from './output.js';
import { StatelessInquirer } from '../service/stateless_inquirer.js';
import { StatelessDeterminism } from '../service/stateless_determinism.js';
import { OutOfAnswerException } from '../service/out_of_answer_error.js';
import type { Question } from '../core/question.js';

class TemplateService {
  private readonly _template: ICyanTemplate;

  constructor(template: ICyanTemplate) {
    this._template = template;
  }

  async template(answer: TemplateInput): Promise<TemplateOutput> {
    let pointer: number = -1;
    const i = new StatelessInquirer(answer.answers, pointer);
    const d = new StatelessDeterminism(answer.deterministicState, pointer);

    try {
      const r = await this._template.template(i, d);
      return {
        data: r,
      } satisfies TemplateFinalOutput;
    } catch (e) {
      if (e instanceof OutOfAnswerException) {
        const q: Question = e.question;
        return {
          deterministicState: d.states,
          question: q,
        } satisfies TemplateQnAOutput;
      }
      throw e;
    }
  }

  async validate(answer: TemplateValidateInput): Promise<string | null> {
    let pointer: number = -1;
    const i = new StatelessInquirer(answer.answers, pointer);
    const d = new StatelessDeterminism(answer.deterministicState, pointer);

    try {
      await this._template.template(i, d);
      throw new Error('Not supposed to reach here for validation!');
    } catch (e) {
      if (e instanceof OutOfAnswerException) {
        const q: Question = e.question;
        const validateResult = q.validate?.(answer.validate);
        return validateResult || null;
      }
      throw e;
    }
  }
}

export { TemplateService };
